### ISSUE TITLE
Fully functional reader macros

### ISSUE THREAD

**Issue by fjarri** [hylang/hy#722]
Currently Hy has some limited support of reader macros (quote, `#` etc). Is it planned (or even wanted) to add full support of reader macros similar to CL's `set-macro-character`? I think it will require either replacing the current lexer+parser pair with a custom reader in Python, or, alternatively, leaving the lexer+parser as the first stage that compiles the actual reader written in Hy.

**Comment by scauligi** [comment]
Bumping this from the grave; I was fooling around and wrote a fairly straightforward reader/parser that reads one character at a time, but my initial prototype seems to actually run a bit _faster_ than the current `rply` lexer/parser.

If there's interest, I can implement the full grammar in this thing and get a PR going; having this would allow us to:
- Get rid of `rply` (which has stagnated)
- Implement constructs like `#_`, `#{...}`, etc. as proper reader macros instead of special-cased grammar rules
- Allow for user-defined reader macros...?(!)
- Have well-bracketed f-strings*, which I could not for the life of me figure out how to implement with `rply`
- Have more control over / have more informative syntax errors
- And also probably still have a (minor) performance improvement over the current lexer/parser

*Well-bracketed f-strings as in: `f"text {(expr 'with "embedded strings" (or f"other {fstrings}"))} more text"`

**Comment by Kodiologist** [comment]
That definitely sounds interesting, although it would probably be better to use `funcparserlib` than write our own parser from scratch.

**Comment by scauligi** [comment]
I tried a parser in `funcparserlib`* (as well as `parsita` and `parsimonious`), and they all end up slower than the current `rply` implementation. I also tried `lark`, which was just about comparable to `rply`.

The custom parser still beats all of these in terms of performance, while also easily allowing for a modifiable reader table for reader macros.

*My funcparserlib implementation is parsing by character instead of tokenizing, since early tokenization is what blocks us from better f-strings, custom readers, etc

----

Results from parsing a ~3100 line file, min of 25 runs reported
```
custom         0.1829 s
lark           0.3227 s
hy (rply)      0.3610 s
parsimonious   0.4295 s
funcparserlib  0.5004 s
parsita        0.5667 s
```
